### PR TITLE
CARBON-15478: Redesign and refactor repo-gen maven goal of carbon-p2-plugin 

### PR DESCRIPTION
**Pull request by dinushab**

### CODE REVIEW

File: carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/install/FeatureInstallResourceBundle.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    private String profile;
+    private URL repository;
+    private List<Feature> features;
+    private boolean deleteOldProfileFiles;
+    private MavenProject project;
+    private P2ApplicationLauncher launcher;
+    private int forkedProcessTimeoutInSeconds;
+    private Log log;
+
+    public String getDestination() {
```
Comment (Kishanthan) [inline carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/install/FeatureInstallResourceBundle.java:43]: Let's add javadocs for all the public methods as a best practice.

File: carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/uninstall/FeatureUnInstaller.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    private void uninstallFeatures(String uninstallUIs) throws MojoFailureException {
+        P2ApplicationLaunchManager launcher = new P2ApplicationLaunchManager(this.launcher);
+        launcher.setWorkingDirectory(project.getBasedir());
+        launcher.setApplicationName(PUBLISHER_APPLICATION);
+        launcher.addArgumentsToUnInstallFeatures(uninstallUIs, destination,
+                profile);
+        launcher.performAction(forkedProcessTimeoutInSeconds);
+    }
+
+    public void setDestination(String destination) {
```
Comment (Kishanthan) [inline carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/uninstall/FeatureUnInstaller.java:66]: Let's add javadocs for all the public methods as a best practice.

File: carbon-p2-plugin/src/main/java/org/wso2/maven/p2/repository/Category.java [hunk 3]
```diff
@@ -15,91 +15,90 @@
...
+        return features;
+    }
 
-	public ArrayList<CatFeature> getFeatures() {
-		return features;
-	}
-	
-    public ArrayList<CatFeature> getProcessedFeatures(MavenProject project, ArtifactFactory artifactFactory, List remoteRepositories, ArtifactRepository localRepository, ArtifactResolver resolver) throws MojoExecutionException{
-        if (processedFeatures != null)
+    public ArrayList<CatFeature> getProcessedFeatures(MavenProject project) throws MojoExecutionException {
```
Comment (Kishanthan) [inline carbon-p2-plugin/src/main/java/org/wso2/maven/p2/repository/Category.java:66]: Add javadocs for all the public methods as a best practice.

File: carbon-p2-plugin/src/main/java/org/wso2/maven/p2/utils/BundleUtils.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        // This method is taken from org.apache.maven.shared.osgi.DefaultMaven2OsgiConverter to get the OSGI version
+        // from a give version artifact version.
+        String osgiVersion;
+
+        // Matcher m = P_VERSION.matcher(version);
+        // if (m.matches()) {
+        // osgiVersion = m.group(1) + "." + m.group(3);
+        // }
+
+        /* TODO need a regexp guru here */
```
Comment (Kishanthan) [inline carbon-p2-plugin/src/main/java/org/wso2/maven/p2/utils/BundleUtils.java:46]: Fix this TODO.

File: carbon-p2-plugin/src/main/java/org/wso2/maven/p2/utils/DependencyResolver.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    /**
+     * Resolve the given project dependencies into CarbonArtifact objects. Dependencies are categorized into
+     * OSGI bundles and Carbon features.
+     *
+     * @param project            MavenProject  Maven Project
+     * @param repositorySystem   RepositorySystem object
+     * @param remoteRepositories collection of remote repositories
+     * @param localRepository    local repository representation
+     * @return Return a List&lt;HashMap&lt;String, CarbonArtifact&gt;&gt;, 1st item being HashMap&lt;String,
```
Comment (Kishanthan) [inline carbon-p2-plugin/src/main/java/org/wso2/maven/p2/utils/DependencyResolver.java:49]: Fix the javadocs alignment issue here.

File: carbon-p2-plugin/src/main/java/org/wso2/maven/p2/utils/BundleUtils.java (hunk 4 above)
Comment (dinushab) [inline carbon-p2-plugin/src/main/java/org/wso2/maven/p2/utils/BundleUtils.java:46]: As I have commented here this method is taken from org.apache.maven.shared.osgi.DefaultMaven2OsgiConverter class. The TODO is in the original source. 

https://svn.apache.org/repos/asf/maven/shared/trunk/maven-osgi/src/main/java/org/apache/maven/shared/osgi/DefaultMaven2OsgiConverter.java

File: carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/generate/Feature.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import org.wso2.maven.p2.beans.CarbonArtifact;
+
+/**
+ * Bean class representing a Feature object provided as an input param to FeatureGenMojo
+ */
+public class Feature extends CarbonArtifact {
+    private String id;
+
+    public String getId() {
+        if(id.endsWith(".feature")) {
```
Comment (Kishanthan) [inline carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/generate/Feature.java:28]: Formatting issue.

File: carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/generate/FeatureResourceBundle.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    private RepositorySystem repositorySystem;
+    private MavenProject project;
+    private ArtifactRepository localRepository;
+    private MavenProjectHelper projectHelper;
+    private List<ArtifactRepository> remoteRepositories;
+
+    private Log log;
+
+    public String getId() {
+        if(id.endsWith(".feature")) {
```
Comment (Nipuni) [inline carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/generate/FeatureResourceBundle.java:62]: formatting issue

Comment (dinushab) [inline carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/generate/FeatureResourceBundle.java:62]: This code is updated and style issues are fixed in a later PR.

**Pull request merged**